akka-stream-tests/src/test/scala/akka/stream/scaladsl/QueueSourceSpec.scala

```diff
@@ -182,6 +182,28 @@ class QueueSourceSpec extends StreamSpec {
       probe.request(6).expectNext(2, 3, 4, 5, 6).expectComplete()
     }
 
+    "allow to wait `n` offer futures in backpressure mode with `n` maxConcurrentPulls" in assertAllStagesStopped {
+      val n = 2
+      val (queue, probe) = Source.queue[Int](5, OverflowStrategy.backpressure, n).toMat(TestSink.probe)(Keep.both).run()
```

Comment:
how is it different from increasing the bufferSize with 2? `Source.queue[Int](7, OverflowStrategy.backpressure)` ?

Reply:
@patriknw `Source.queue[Int](m, OverflowStrategy.backpressure, n)` can be used in places with limited by n concurrency. Here is an example of desired code for the similar feature for `Sink.queue`: https://github.com/akka/akka/issues/27361#issuecomment-511799230

Reply:
ok, so you know that there will only be `n` producers emitting to the `queue`, and each producer will wait for its offer result to complete before emitting next element. Sounds good to be able to support that. I'll have to think some more about this next week.

Vague thought right now is that maybe we could have a new backpressure strategy which will not complete the Future[QueueOfferResult] until the element has been emitted downstreams. Then the ordinary buffer can hold all the pending offers.

Reply:
yes, exactly

using the ordinary buffer for pending offer will make API a bit clunky, because of mixing of definitions. But mb it will be better to remove ordinary buffer at all and make backpressure as default overflow strategy, everyone is allowed to compose the queue with downstream buffer.

here is the similar feature request for `Sink.queue`: https://github.com/akka/akka/issues/28270